### PR TITLE
[MV-419] Port: MSELoss kernel family to MIOpen

### PR DESCRIPTION
**Pull request by o2buzzle**
This PR ports the `MSELoss` family of loss function to MIOpen:
- `MSELoss`
- `MSELossUnreduced`

Performance measurements seems to suggest that in general we're performing better than ROCm on forward, reduced operation (mostly thanks to parallel reduction). 

### Sample performance measurements

<details>
<summary>float32</summary>

|op_name|dtype  |Size-reduction         |contiguous   |model    |direction|rocm kernel time|miopen kernel time|miopendriver cmdline                        |rocm/miopen     |
|-------|-------|-----------------------|-------------|---------|---------|----------------|------------------|--------------------------------------------|----------------|
|MSELoss|float32|[32 80]-sum            |noncontiguous|tacotron2|bwd      |20784           |11804             |mseloss -I 32x80 -r sum -Z 0_bwd            |1.76075906472382|
|MSELoss|float32|[25 100 100]-mean      |noncontiguous|random-md|fwd      |27568           |16551             |mseloss -I 25x100x100 -r mean -Z 0          |1.66563953839647|
|MSELoss|float32|[25 300]-sum           |noncontiguous|random-md|bwd      |15888           |10204             |mseloss -I 25x300 -r sum -Z 0_bwd           |1.55703645629165|
|MSELoss|float32|[10 10 50 100 100]-none|noncontiguous|random-lg|bwd      |14407533        |9301430           |mseloss -I 10x10x50x100x100 -r none -Z 0_bwd|1.54895892352036|
|MSELoss|float32|[32 80]-mean           |contiguous   |tacotron2|fwd      |16544           |10791             |mseloss -I 32x80 -r mean -Z 1_contig        |1.53312945973496|
|MSELoss|float32|[25 300]-mean          |noncontiguous|random-md|bwd      |15632           |10400             |mseloss -I 25x300 -r mean -Z 0_bwd          |1.50307692307692|
|MSELoss|float32|[32 80]-none           |noncontiguous|tacotron2|bwd      |17760           |12089             |mseloss -I 32x80 -r none -Z 0_bwd           |1.46910414426338|
|MSELoss|float32|[32 80]-sum            |contiguous   |tacotron2|fwd      |15952           |10862             |mseloss -I 32x80 -r sum -Z 1_contig         |1.46860614988032|
|MSELoss|float32|[10 100 100 100]-none  |noncontiguous|random-lg|bwd      |2423129         |1661950           |mseloss -I 10x100x100x100 -r none -Z 0_bwd  |1.45800355004663|
|MSELoss|float32|[2000 3000]-none       |noncontiguous|random-lg|bwd      |547674          |376328            |mseloss -I 2000x3000 -r none -Z 0_bwd       |1.45531026126145|
|MSELoss|float32|[25 100 100]-sum       |contiguous   |random-md|fwd      |23745           |16676             |mseloss -I 25x100x100 -r sum -Z 1_contig    |1.42390261453586|
|MSELoss|float32|[25 100 100]-mean      |contiguous   |random-md|fwd      |23760           |16711             |mseloss -I 25x100x100 -r mean -Z 1_contig   |1.42181796421519|
|MSELoss|float32|[2000 3000]-sum        |contiguous   |random-lg|fwd      |96514           |68268             |mseloss -I 2000x3000 -r sum -Z 1_contig     |1.41375168453741|
|MSELoss|float32|[1000000]-mean         |noncontiguous|random-lg|fwd      |33825           |24231             |mseloss -I 1000000 -r mean -Z 0             |1.39593908629442|
|MSELoss|float32|[32 80 870]-sum        |contiguous   |t5       |fwd      |48673           |34880             |mseloss -I 32x80x870 -r sum -Z 1_contig     |1.39544151376147|
|MSELoss|float32|[100 20 20 20]-none    |noncontiguous|random-md|bwd      |83490           |59877             |mseloss -I 100x20x20x20 -r none -Z 0_bwd    |1.39435843479132|
|MSELoss|float32|[32 80 870]-mean       |contiguous   |t5       |fwd      |48273           |34934             |mseloss -I 32x80x870 -r mean -Z 1_contig    |1.38183431613901|
|MSELoss|float32|[100 10 10 10 10]-none |noncontiguous|random-md|bwd      |98626           |71770             |mseloss -I 100x10x10x10x10 -r none -Z 0_bwd |1.37419534624495|
|MSELoss|float32|[128 80 870]-sum       |contiguous   |tacotron2|fwd      |129890          |94687             |mseloss -I 128x80x870 -r sum -Z 1_contig    |1.37178282129543|
|MSELoss|float32|[2000 3000]-mean       |contiguous   |random-lg|fwd      |94226           |68694             |mseloss -I 2000x3000 -r mean -Z 1_contig    |1.37167729350453|

</details>

<details>
<summary>float16</summary>

|op_name|dtype  |Size-reduction         |contiguous   |model    |direction|rocm kernel time|miopen kernel time|miopendriver cmdline                        |rocm/miopen     |
|-------|-------|-----------------------|-------------|---------|---------|----------------|------------------|--------------------------------------------|----------------|
|MSELoss|float16|[25 100 100]-mean      |noncontiguous|random-md|fwd      |30241           |16124             |mselossfp16 -I 25x100x100 -r mean -Z 0      |1.87552716447532|
|MSELoss|float16|[25 300]-none          |noncontiguous|random-md|bwd      |18624           |10080             |mselossfp16 -I 25x300 -r none -Z 0_bwd      |1.84761904761905|
|MSELoss|float16|[25 100 100]-sum       |noncontiguous|random-md|fwd      |28785           |16160             |mselossfp16 -I 25x100x100 -r sum -Z 0       |1.78125         |
|MSELoss|float16|[32 80]-mean           |contiguous   |tacotron2|fwd      |16880           |10506             |mselossfp16 -I 32x80 -r mean -Z 1_contig    |1.6067009328003 |
|MSELoss|float16|[32 80]-sum            |contiguous   |tacotron2|fwd      |17008           |10649             |mselossfp16 -I 32x80 -r sum -Z 1_contig     |1.59714527185651|
|MSELoss|float16|[25 300]-sum           |noncontiguous|random-md|bwd      |16368           |10275             |mselossfp16 -I 25x300 -r sum -Z 0_bwd       |1.59299270072993|
|MSELoss|float16|[32 80]-sum            |noncontiguous|tacotron2|bwd      |19920           |12515             |mselossfp16 -I 32x80 -r sum -Z 0_bwd        |1.59168997203356|
|MSELoss|float16|[25 300]-mean          |noncontiguous|random-md|bwd      |16448           |10453             |mselossfp16 -I 25x300 -r mean -Z 0_bwd      |1.5735195637616 |
|MSELoss|float16|[32 80]-none           |noncontiguous|tacotron2|bwd      |18128           |12035             |mselossfp16 -I 32x80 -r none -Z 0_bwd       |1.50627336933943|
|MSELoss|float16|[25 100 100]-sum       |contiguous   |random-md|fwd      |23392           |15911             |mselossfp16 -I 25x100x100 -r sum -Z 1_contig|1.47017786437056|
|MSELoss|float16|[32 80]-mean           |noncontiguous|tacotron2|bwd      |17344           |12125             |mselossfp16 -I 32x80 -r mean -Z 0_bwd       |1.43043298969072|
|MSELoss|float16|[10 10 50 100 100]-none|noncontiguous|random-lg|bwd      |11586974        |8183960           |mselossfp16 -I 10x10x50x100x100 -r none -Z 0_bwd|1.41581508218515|
|MSELoss|float16|[25 100 100]-mean      |contiguous   |random-md|fwd      |22784           |16107             |mselossfp16 -I 25x100x100 -r mean -Z 1_contig|1.41454026199789|
|MSELoss|float16|[25 100 100]-none      |noncontiguous|random-md|bwd      |30769           |22134             |mselossfp16 -I 25x100x100 -r none -Z 0_bwd  |1.39012379145206|
|MSELoss|float16|[100 20 20 20]-none    |noncontiguous|random-md|bwd      |75393           |57352             |mselossfp16 -I 100x20x20x20 -r none -Z 0_bwd|1.31456618775282|
|MSELoss|float16|[100 10 10 10 10]-none |noncontiguous|random-md|bwd      |89410           |68943             |mselossfp16 -I 100x10x10x10x10 -r none -Z 0_bwd|1.29686842754159|
|MSELoss|float16|[2000 3000]-none       |noncontiguous|random-lg|bwd      |463896          |360150            |mselossfp16 -I 2000x3000 -r none -Z 0_bwd   |1.28806330695544|
|MSELoss|float16|[10 100 100 100]-none  |noncontiguous|random-lg|bwd      |1869280         |1504310           |mselossfp16 -I 10x100x100x100 -r none -Z 0_bwd|1.24261621607249|
|MSELoss|float16|[25 100 100]-mean      |noncontiguous|random-md|bwd      |26752           |21618             |mselossfp16 -I 25x100x100 -r mean -Z 0_bwd  |1.23748727911925|
|MSELoss|float16|[1000000]-mean         |contiguous   |random-lg|fwd      |29456           |24249             |mselossfp16 -I 1000000 -r mean -Z 1_contig  |1.21473050435069|

</details>

<details>
<summary>bfloat16</summary>

|op_name|dtype  |Size-reduction         |contiguous   |model    |direction|rocm kernel time|miopen kernel time|miopendriver cmdline                        |rocm/miopen     |
|-------|-------|-----------------------|-------------|---------|---------|----------------|------------------|--------------------------------------------|----------------|
|MSELoss|bfloat16|[100 20 20 20]-sum     |noncontiguous|random-md|fwd      |43889           |22329             |mselossbfp16 -I 100x20x20x20 -r sum -Z 0    |1.96556048188454|
|MSELoss|bfloat16|[25 300]-sum           |contiguous   |random-md|fwd      |19904           |10169             |mselossbfp16 -I 25x300 -r sum -Z 1_contig   |1.95732127052808|
|MSELoss|bfloat16|[100 10 10 10 10]-mean |noncontiguous|random-md|fwd      |48049           |24693             |mselossbfp16 -I 100x10x10x10x10 -r mean -Z 0|1.94585510063581|
|MSELoss|bfloat16|[10000]-mean           |noncontiguous|random-md|fwd      |19664           |10151             |mselossbfp16 -I 10000 -r mean -Z 0          |1.93714904935474|
|MSELoss|bfloat16|[100 20 20 20]-mean    |noncontiguous|random-md|fwd      |42273           |22347             |mselossbfp16 -I 100x20x20x20 -r mean -Z 0   |1.89166331051148|
|MSELoss|bfloat16|[25 300]-none          |noncontiguous|random-md|bwd      |19216           |10293             |mselossbfp16 -I 25x300 -r none -Z 0_bwd     |1.86689983483921|
|MSELoss|bfloat16|[25 100 100]-sum       |noncontiguous|random-md|fwd      |29441           |16302             |mselossbfp16 -I 25x100x100 -r sum -Z 0      |1.80597472702736|
|MSELoss|bfloat16|[25 100 100]-mean      |noncontiguous|random-md|fwd      |28848           |16107             |mselossbfp16 -I 25x100x100 -r mean -Z 0     |1.79102253678525|
|MSELoss|bfloat16|[32 80]-mean           |contiguous   |tacotron2|fwd      |17456           |10240             |mselossbfp16 -I 32x80 -r mean -Z 1_contig   |1.7046875       |
|MSELoss|bfloat16|[32 80]-sum            |contiguous   |tacotron2|fwd      |17616           |10507             |mselossbfp16 -I 32x80 -r sum -Z 1_contig    |1.67659655467783|
|MSELoss|bfloat16|[32 80]-sum            |noncontiguous|tacotron2|bwd      |21296           |13013             |mselossbfp16 -I 32x80 -r sum -Z 0_bwd       |1.63651732882502|
|MSELoss|bfloat16|[25 100 100]-sum       |contiguous   |random-md|fwd      |24832           |15982             |mselossbfp16 -I 25x100x100 -r sum -Z 1_contig|1.55374796646227|
|MSELoss|bfloat16|[25 100 100]-mean      |contiguous   |random-md|fwd      |24592           |15911             |mselossbfp16 -I 25x100x100 -r mean -Z 1_contig|1.5455973854566 |
|MSELoss|bfloat16|[10 100 100 100]-none  |noncontiguous|random-lg|bwd      |1945089         |1296260           |mselossbfp16 -I 10x100x100x100 -r none -Z 0_bwd|1.50053924367025|
|MSELoss|bfloat16|[10 10 50 100 100]-none|noncontiguous|random-lg|bwd      |11904466        |8223780           |mselossbfp16 -I 10x10x50x100x100 -r none -Z 0_bwd|1.44756620434885|
|MSELoss|bfloat16|[25 300]-sum           |noncontiguous|random-md|bwd      |15104           |10595             |mselossbfp16 -I 25x300 -r sum -Z 0_bwd      |1.42557810287872|
|MSELoss|bfloat16|[25 300]-mean          |noncontiguous|random-md|bwd      |14992           |10649             |mselossbfp16 -I 25x300 -r mean -Z 0_bwd     |1.40783172128838|
|MSELoss|bfloat16|[25 100 100]-none      |noncontiguous|random-md|bwd      |31489           |22685             |mselossbfp16 -I 25x100x100 -r none -Z 0_bwd |1.38809786202336|
|MSELoss|bfloat16|[32 80]-none           |noncontiguous|tacotron2|bwd      |17792           |13155             |mselossbfp16 -I 32x80 -r none -Z 0_bwd      |1.35248954770049|
|MSELoss|bfloat16|[32 80]-mean           |noncontiguous|tacotron2|bwd      |16928           |12835             |mselossbfp16 -I 32x80 -r mean -Z 0_bwd      |1.3188936501753 |

</details>

### Average performance
- `MSELoss`

|dtype|forward|backward|
|-----|-------|--------|
|float32|2.52|0.74|
|float16|2.08|0.66|
|bfloat16|2.12|0.66|

- `MSELossUnreduced`

|dtype|forward|backward|
|-----|-------|--------|
|float32|0.53|0.92|
|float16|0.45|0.82|
|bfloat16|0.49|0.86|

### CODE REVIEW

**Comment by et16kr** [comment]
In Average performance, only reduced forward is fast. 
However, the performance sample shows that backward and unreduced are also fast. 
Which one is correct?

And.. 
If only reduced forward is faster than rocm pytorch, 
unreduced and backward should be removed and and filtered out in IsApplicable().

**Comment by o2buzzle** [comment]
@et16kr You can check the attached documents in the Jira ticket mentioned in the PR for more details, basically only **some** of those ops even managed > 0.8, let alone 1.2. There are just so few cases that passes the threshold. The ones that appears faster in my measurement seems to only be 2d non-contiguous tensors.

Also, I noticed that some other PR doesn't provide any benchmark sample at all, only average, so I presumed that it should only be a sample of what was tested, not everything that was tested.

And about Unreduced + Backward, those will be filtered out, since their `IsImprovementOverROCm` is a static `return false;`, with accompanying comments as to why that is. If it is necessary I can completely remove those sections. Those are only provided in case you want to re-run a different benchmark set than mine, then making them available is easier than having to roll back commits.

**Comment by o2buzzle** [comment]
@et16kr I've added this specific case that `MSELossBackward` wins (2d, non-contiguous tensor input) in commit https://github.com/moreh-dev/MIOpen/pull/18/commits/7b5af98be3654087fd80540e846a2e416d8fc820. The performance average for this specific case is as following:

|dtype|MSELossBackward|
|--------|---------------------------|
|fp32  |1.46 |
|fp16  |1.42 |
|bfp16|1.38|

**Comment by et16kr** [comment]
An error occur in driver. Please check for error.
```
# bin/MIOpenDriver mseloss -I 25x100x100 -Z 1 -r sum
MIOpenDriver mseloss -I 25x100x100 -Z 1 -r sum
PRNG seed: 12345678
Created workspace with size 1003908
Error: Forward CPU and GPU mismatch
output[0] = 41651.3 != 41648.4
```

**Comment by et16kr** [comment]
- miopenMSELossForwardUnreduced() and miopenMSELossBackwardUnreduced() do not work. 
You should not add non-working APIs to MIOpen.
- But if you found some quick parts, please keep the code and modify IsImprovementOverROCm(). (like reduced backward)
- There is only reduced forward in gtest. If reduced backward is added, please add a test.
- Please remove unused cpu code from cpu_mseloss.cpp.

File: include/miopen/miopen.h [hunk 1]
```diff
@@ -6577,11 +6577,139 @@ MIOPEN_EXPORT miopenStatus_t miopenBackendDestroyDescriptor(miopenBackendDescrip
...
+ *
+ * MSELoss forward, reduced function
+ * @param  [in]  handle         An instance of miopenHandle_t
+ * @param  [in]  xDesc          Input Tensor descriptor
+ * @param  [in]  yDesc          Target Tensor descriptor
+ * @param  [in]  zDesc          Output Tensor descriptor
+ * @param  [in]  x              Pointer to input tensor data
+ * @param  [in]  y              Pointer to target tensor data
+ * @param  [out] z              Pointer to the output tensor data
+ * @param  [out]  workspace      Pointer to the workspace area (for reduction)
```
Comment (et16kr) [inline include/miopen/miopen.h]: Please match the code indentation.

File: include/miopen/miopen.h [hunk 2]
```diff
@@ -512,7 +512,7 @@ typedef enum
     miopenActivationABS      = 5, /*!< Absolute value \f$abs(x)\f$ */
     miopenActivationPOWER = 6, /*!< Scaled and shifted power \f$(\alpha + \beta * x)^{gamma}\f$ */
     miopenActivationCLIPPEDRELU =
-        7, /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
+        7,                     /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
```
Comment (et16kr) [inline include/miopen/miopen.h:515]: Please revert code. An error occurs in clang-format.

File: driver/mseloss_driver.hpp [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+    }
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int MSELossDriver<Tgpu, Tref>::VerifyBackward()
+{
+    RunBackwardCPU();
+
+    for(size_t i = 0; i < input_grad.size(); i++)
```
Comment (et16kr) [inline driver/mseloss_driver.hpp]: Please use miopen::rms_range.
Or, if you have a reason for doing this, please tell me.

File: driver/mseloss_driver.hpp [hunk 4]
```diff
@@ -0,0 +1,732 @@
...
+};
+
+template <typename Tgpu, typename Tref>
+int MSELossDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "1", "Run only Forward Cat (Default=1)", "int");
+    inflags.AddInputFlag("contiguous", 'Z', "0", "Use Contiguous Tensors", "int");
+    inflags.AddTensorFlag("in_tensors", 'I', "1", "Input Tensors");
+    inflags.AddInputFlag("divisor", 'D', "1", "Divisor", "float");
+    inflags.AddInputFlag("reduction", 'r', "custom", "Reduction", "string");
```
Comment (et16kr) [inline driver/mseloss_driver.hpp]: Please indicate "none", "sum", "mean" in the reduction of MIOpenDriver mseloss --help.

**Comment by o2buzzle** [comment]
> An error occur in driver. Please check for error.
> 
> ```
> # bin/MIOpenDriver mseloss -I 25x100x100 -Z 1 -r sum
> MIOpenDriver mseloss -I 25x100x100 -Z 1 -r sum
> PRNG seed: 12345678
> Created workspace with size 1003908
> Error: Forward CPU and GPU mismatch
> output[0] = 41651.3 != 41648.4
> ```

I believe this is an accumulated precision error between CPU and GPU floating point calculations. This is because if you do the same operation but unreduced (enable it by setting the `IsImprovementOverROCm` to return `true`), it returns correctly.

```
MIOpenDriver mseloss -I 25x100x100 -Z 1 -r none
MIOpen(HIP): Info [get_device_name] Raw device name: gfx90a:sramecc+:xnack-
MIOpen(HIP): Info [Handle] stream: 0x22c5100, device_id: 0
PRNG seed: 12345678
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, HIP version 6.0.24015, MIOpen version 3.1.0.7b5af98be-dirty
MIOpen(HIP): Info [KernDb] database not present
MIOpen(HIP): Info [PrintVersion] HIPRTC v.9.0
MIOpen(HIP): Info [KernDb] database not present
Success: Forward CPU and GPU match
```

Even with the `rms_range` method, this results in a difference of 7e-5 

I’ll see if there is a way I can match them exactly, but I have been running into it here and there while testing this.

File: include/miopen/miopen.h (hunk 2 above)
Comment (o2buzzle) [reply include/miopen/miopen.h:515]: I’ve manually removed the changes

File: include/miopen/miopen.h (hunk 1 above)
Comment (o2buzzle) [reply include/miopen/miopen.h]: I’ve updated the indentation to match others

File: driver/mseloss_driver.hpp (hunk 4 above)
Comment (o2buzzle) [reply driver/mseloss_driver.hpp]: I’ve added the descriptions

File: driver/mseloss_driver.hpp (hunk 3 above)
Comment (o2buzzle) [reply driver/mseloss_driver.hpp]: It was associated with the precision issue. I’ve changed to `rms_range`

**Comment by o2buzzle** [comment]
@et16kr  I've managed to reduce the error down low enough that `rms_range` seems happy with a tolerance found in other drivers. Please check if any other issues are present

**Comment by o2buzzle** [comment]
> * miopenMSELossForwardUnreduced() and miopenMSELossBackwardUnreduced() do not work.
>       You should not add non-working APIs to MIOpen.
> 
>     * But if you found some quick parts, please keep the code and modify IsImprovementOverROCm(). (like reduced backward)
> 
>     * There is only reduced forward in gtest. If reduced backward is added, please add a test.
> 
>     * Please remove unused cpu code from cpu_mseloss.cpp.

Most of the changes here has been made in https://github.com/moreh-dev/MIOpen/pull/18/commits/e4dfd03aeaeb94e794ffd7a508bb71a17dff7cbb and https://github.com/moreh-dev/MIOpen/pull/18/commits/43185b8bc2a2c3785d62f38fb9da07834a9303ab

**Comment by et16kr** [comment]
Good work! Thank you for your hard work.

**Pull request closed without merging**